ТЗ 11 Добавление Базы Данных

RomanAntipenko opened the pull request:
feat: добавь новую реализацию новых хранилищ в виде БД
feat: добавь новые функции у фильмов в виде MPArating, и Genre
feat: реализуй для всего этого сервисы и контролерры т.к. появились новые эндпоинты
feat: добавь скрипты для инициализации бд
feat: добавь для новых контроллеров тесты

RomanAntipenko committed:
feat: добавь новую реализацию новых хранилищ в виде БД
feat: добавь новые функции у фильмов в виде MPArating, и Genre
feat: реализуй для всего этого сервисы и контролерры т.к. появились новые эндпоинты
feat: добавь скрипты для инициализации бд
feat: добавь для новых контроллеров тесты

RomanAntipenko committed:
feat: добавь новую реализацию новых хранилищ в виде БД
feat: добавь новые функции у фильмов в виде MPArating, и Genre
feat: реализуй для всего этого сервисы и контролерры т.к. появились новые эндпоинтыfeat: добавь скрипты для инициализации бд
feat: добавь для новых контроллеров тесты
fix: CheckStyle...

RomanAntipenko committed:
feat: добавь новую реализацию новых хранилищ в виде БД
feat: добавь новые функции у фильмов в виде MPArating, и Genre
feat: реализуй для всего этого сервисы и контролерры т.к. появились новые эндпоинтыfeat: добавь скрипты для инициализации бд
feat: добавь для новых контроллеров тесты
fix: CheckStyle...x2

stfaustoff commented on src/main/java/ru/yandex/practicum/filmorate/dao/GenreDao.java:
```diff
@@ -0,0 +1,70 @@
+
+
+    public Genre getGenreByGenreId(int genreId) {
+        if (isValidId(genreId)) {
```
А есть ли смысл заранее запрашивать кол-во записей по идентификатору, если мы дальше делаем ровно такой же запрос. Получится что мы делаем на 1 запрос больше, хотя в итоге результат получим тот же самый.

stfaustoff commented on src/main/java/ru/yandex/practicum/filmorate/dao/GenreDao.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public List<Genre> getGenreListByFilmId(long filmId) {
+        String sql = "SELECT fg.genre_id, g.genre_name FROM public.films AS f " +
```
Может лучше делать From genre, чем film. Так будет логичнее.

stfaustoff commented on src/main/java/ru/yandex/practicum/filmorate/dao/MpaDao.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public Mpa getMpaByMpaId(int mpaId) {
+        if (isValidId(mpaId)) {
```
Тут вопрос с валидацией аналогично актуален.

stfaustoff commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDBStorage.java:
```diff
@@ -0,0 +1,164 @@
+        if (filmStorageValidation(filmValid)) {
+            String sql = "SELECT * FROM public.films WHERE film_id = ?";
+            /*return jdbcTemplate.queryForObject(sql, (rs, rowNum)-> makeUser(rs), id);*/
+            Film film = jdbcTemplate.queryForObject(sql, this::makeFilm, id);
```
Получение одной записи из БД теперь сопряжено с большим кол-вом запросов.
Здесь нужно перейти на Join, как минимум по mpa.

stfaustoff commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,63 @@
+CREATE TABLE IF NOT EXISTS PUBLIC.USERS (
+	USER_ID INTEGER GENERATED BY DEFAULT AS IDENTITY,
+	USER_NAME VARCHAR_IGNORECASE(50),
+	USER_LOGIN VARCHAR_IGNORECASE(15) NOT NULL,
```
_Предложение_
Можно добавить уникальный индекс, чтобы и БД отслеживало отсутствие повторяющихся данных в поле

stfaustoff commented on tables.png:
Диаграмма реализована верно, осталось только добавить ее в Readme

RomanAntipenko committed:
fix: убери лишнюю валидацию из МПА и GENRE проверку организуй через sqlForRowSet чтобы избежать лишних запросов
fix: обьедени в нескольких методах запросы к МПА и Фильмам в один запрос
feat: добавь уникальные индексы в USER_LOGIN для проверки в дб на уникальность логинов
style: убери закомментированные куски кода
feat: добавь схему в ReadMe
fix: getGenreListByFilmId подправь JOIN
